### PR TITLE
Graphene Evaluation

### PR DESCRIPTION
**Pull request by gtalarico**
So far seems to work well - would be so happy to strip out all those views and serializers.
~Need to test some more complex relationships and queries see how this flexes.~

### Todo
- [X] Test Mutations/Write
- [X] Test Filter by multiple hashtags (eg. `companies/?hashtags=web,api` -> `{ companies (hashtags: ["a", "b"] }`)
- [ ] Test Pagination

...

image

PS: GraphiQl working 👍 

image

--- 

### Mutations

**Mutations**
I evaluated graphene's mutations and DRF SerializerMutation.
I think SerializerMutation creates too tight a coupling between model and serialization.
The regular mutation seems adequate enough.

**Auth/Middleware**
While we don't have a full strategy for auth and permissions, the attached middleware examples shows a way to implement something similar to our most common use case (POST/mutations require auth - clap, comment, create, etc).

TODO: not sure where how permission like `isEditor` would be enforced

image

### CODE REVIEW

**Comment by danrocha** [comment]
Does Django's plugin provide a GraphiQL interface to play with the API? 
If so, it is usually more useful than Postman for that... especially if it has an Explorer. Also, it comes with Documentation built in.

Otherwise, you can also try to play with it here https://graphiql-online.com/

**Review by danrocha — APPROVED** [review]
Oh, I see the GraphiQL = true there... nice.

**Comment by gtalarico** [comment]
yep! didn't test the UI interface but should be working!
FYI, what's left to do here is to try implementing some of the mutations and see how easily we can replace the existing views.

I am also curios, on the front end, would you suggest just using regular fetch requests with gql payload? or use a client like apollo?

**Comment by danrocha** [comment]
Apollo is great, integrated well with Vue, and has its own cache, checks for repeated queries, integrates well with vue, etc etc. etc. there are many more features we'll prob never use though...

However, going in the "simple fetch" direction, there is another interesting library we can check: [Villus](https://github.com/logaretm/villus). I love its simplicity and tried using it a while ago, but it was too early in their development and a bit too buggy for production. They are approaching v1.0 soon though and getting commits every day.

File: api/gql_schema.py [hunk 1]
```diff
@@ -76,6 +79,43 @@ class Meta:
         fields = "__all__"
 
 
+class CreatePostWithSerializer(SerializerMutation):
```
Comment (gtalarico) [inline api/gql_schema.py:82]: Did not like `SerializerMutation` - too much magic, hard to understand and modify behaviour

File: api/gql_schema.py [hunk 2]
```diff
@@ -76,6 +79,43 @@ class Meta:
...
+    def get_serializer_kwargs(cls, root, info, **input):
+        # Gets messy
+        input["profile"] = info.context.user.profile
+        kwargs = super(CreatePostWithSerializer, cls).get_serializer_kwargs(
+            root, info, **input
+        )
+        return kwargs
+
+
+class CreatePost(graphene.Mutation):
```
Comment (gtalarico) [inline api/gql_schema.py:98]: this seems simple enough and would be really easy to migrate from our current viewsets. 
The only thing is DRF model serializers have better built validation that might need to happen manually here, maybe pre-service calls

**Comment by gtalarico** [comment]
* Refactored into sep. files
* Added all/almost all required Queries
* Added a few query filters/search

```
query {
    commentsByThreadId (threadId: 1) {
        text
        clappers {
            name
        }
        profile {
            name
        }
    }
    commentsByParentId (parentId: 1) {
        text
    }
    companies (hashtagNames: "housing") {
        name 
        description
        clapCount
        threadSize
        articles {
            url
        }
        clappers {
            name
        }
        thread { 
            id
        }
        hashtags {
            slug
        }
    }
    posts (hashtagNames: ["social"]) {
        title,
        hashtags {
            slug
        }
    }  
    companyBySlug (slug: "abodu")  {
    name 
    description
    twitterHandle
    articles {
        url
    }
    }
}
```

**Comment by gtalarico** [comment]
Looks like django graphene can handle some of the filtering automagically, could simplify things a bit:
https://github.com/graphql-python/graphene-django/blob/master/examples/cookbook/cookbook/ingredients/schema.py#L31

Also should try other mutation base classes instead of creating them manually:
https://docs.graphene-python.org/projects/django/en/latest/mutations/

**Pull request closed without merging**